### PR TITLE
Remove Source.behavior, fix bug in depfile invalidation

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Remove Source.behavior and all custom asset behavior. This is now replaced with a single helper method that returns a depfile.

Fixes a bug in depfile invalidation. IF a depfile already existed, we would not recompute the outputs after running it. Thus, changes in the outputs expressed in the depfile would not be picked up until a subsequent run. This delayed the removal of stale files.

See the test case that demonstrates the issue

The test skip is due to a bug in package:file https://github.com/google/file.dart/issues/131

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I have a fix pending that allows me to remove the skip in https://github.com/google/file.dart/pull/132

File: packages/flutter_tools/lib/src/build_system/build_system.dart [hunk 1]
```diff
@@ -526,17 +526,17 @@ class _BuildInstance {
       await node.target.build(environment);
       printTrace('${node.target.name}: Complete');
 
-      // If we were missing the depfile, resolve files after executing the
+      node.inputs.clear();
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/build_system.dart]: Maybe:
```dart
node.inputs
  ..clear()
  ..addAll(...);
node.outputs
  ..clear()
  ..addAll(...);
```

File: packages/flutter_tools/lib/src/build_system/depfile.dart [hunk 2]
```diff
@@ -58,6 +58,9 @@ class Depfile {
   ///
   /// If either [inputs] or [outputs] is empty, does not write to the file.
   void writeToFile(File depfile) {
+    if (depfile.existsSync()) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/depfile.dart]: The `writeAsStringSync()` below will truncate the file, so I guess this delete is needed for the case where `inputs` or `outputs` is empty. Maybe add a comment to that effect, or move the delete under the if for the inputs/outputs.

File: packages/flutter_tools/lib/src/build_system/targets/assets.dart [hunk 3]
```diff
@@ -10,81 +10,48 @@ import '../../devfs.dart';
...
-      results.add(file);
-    }
-    return results;
-  }
+import '../depfile.dart';
+
+/// A helper function to copy an asset bundle into an [environment]'s output
+/// directory.
+///
+/// [pathSuffix] may be optionally provided to add additional paths to the
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/build_system/targets/assets.dart]: It looks like this parameter no longer eixsts.

File: packages/flutter_tools/lib/src/build_system/build_system.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/build_system.dart]: Done

File: packages/flutter_tools/lib/src/build_system/depfile.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/depfile.dart]: Moved under the if and added more to the method's doc comment

File: packages/flutter_tools/lib/src/build_system/targets/assets.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/targets/assets.dart]: Removed

File: packages/flutter_tools/test/general.shard/build_system/targets/assets_test.dart [hunk 4]
```diff
@@ -65,7 +67,7 @@ flutter:
 
     // See https://github.com/flutter/flutter/issues/35293
     expect(fs.file(fs.path.join(environment.buildDir.path, 'flutter_assets', 'assets/foo/bar.png')).existsSync(), false);
-  }));
+  }), skip: Platform.isWindows); // See https://github.com/google/file.dart/issues/131
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/build_system/targets/assets_test.dart]: Does this imply that stale files will be left in the build directory on Windows? Could this result in an incorrect build?

**Review by zanderso — APPROVED** [review]
lgtm w/ question

File: packages/flutter_tools/test/general.shard/build_system/targets/assets_test.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/build_system/targets/assets_test.dart]: I could not reproduce this on a real filesystem, which leads me to believe it is only an issue with the memory filesystem.

**Pull request merged**